[benchmark] RomanNumbers Redux

palimondo opened the pull request:
During the Swift Benchmark Suite clean-up, I wasn't able to easily lower the workload and apply `legacyFactor` to the `RomanNumbers` benchmark. Upon closer inspection, I've realized there was a bug in how the `position` was always reset to `0` after `continue outer`.

This PR rewrites the benchmark by swapping the `while` and `for-in` loops into more natural order, which also eliminates the need for using `position` at all. This is based on examples from [Rosetta Code](https://www.rosettacode.org/wiki/Roman_numerals/Decode#Swift).

The benchmark name is changed to `Roman.Substring.startsWith.dropFirst` because of the bug-fix and lowered workload (to run in less than 1000 μs).

palimondo committed:
[benchmark] RomanNumbers Redux

Rewrite of the `RomanNumbers` benchmark that fixes error in the `position` which was always reset to 0 after `continue outer`, by swapping the `while` and `for-in` loops into more natural order, which eliminates the need for using `position` at all. This is based on examples from [Rosetta Code](https://www.rosettacode.org/wiki/Roman_numerals/Decode#Swift).

The coverage has been expanded to measure performance of `String`/`Substring`/`UTF8View` with very short string arguments and methods: `hasPrefix`/`starts(with:)`, `removeFirst`/`dropFirst` and `String.append()`, as well as functional variants of the fastest imperative version.

The `Roman.Substring.startsWith.dropFirst` variant tests the same code paths as the original `RomanNumbers` benchmark, but the name is changed because of the bug-fix and lowered workload (to run in less than 1000 μs).

For comparison, there's one extra variant with character based parsing algorithm: `Roman.DictCharInt.map.reduce`.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@lorentey @airspeedswift @milseman Please review 🙏.

Michael, are there any plans to improve `Substring` performance for small strings? I'm surprised it's consistently slower that using a string directly. Also I find the performance regression of `removeFirst` vs. `dropFirst`+`String.init` puzzling. `String.hasPrefix` shows 2x advantage vs. `starts(with:)` thanks to `String`'s fast path in [`StringLegacy`](https://github.com/apple/swift/blob/870dae52040f52f2ff34845d008f60fef4dc2e4d/stdlib/public/core/StringLegacy.swift#L124). Can we do something similar for the `Sequence`/`Collection` API access to string guts?

The FP-style variants are being disadvantaged because something (ARC?) is preventing tail call optimization of the recursive function which replaced the while loop… Is there nothing that can be done there?

airspeedswift commented:
I'm really not seeing the benefit of adding so many near-identical versions of this benchmark.

palimondo commented:
These benchmarks demonstrate relative performance between the `String` and `Substring` and their various method combinations which I found surprising. Are you saying that the issues I've [raised about above](https://github.com/apple/swift/pull/22227#issuecomment-458722271) are obvious, well known or otherwise uninteresting?

palimondo commented:
To avoid getting stuck on the benchmarking coverage debate, which would further delay the Swift Benchmark Suite cleanup, I'll just do the fixed `Roman.Substring.startsWith.dropFirst` for now…

@milseman Could you please chime in on the [performance surprises above](https://github.com/apple/swift/pull/22227#issuecomment-458722271)?
@airspeedswift What considers a beneficial benchmark coverage?

palimondo committed:
[benchmark] RomanNumeral pared back down

Remove the extended benchmark coverage to leave only bug-fixed version of `RomanNumeral` benchmark with the new name `Roman.Substring.startsWith.dropFirst`.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
I'm saving here the original PR description for 0edd560, in case we decide to expand the benchmark coverage later on...

----

During the Swift Benchmark Suite clean-up, I wasn't able to easily lower the workload and apply `legacyFactor` to the `RomanNumbers` benchmark. Upon closer inspection, I've realized there was a bug in how the `position` was always reset to `0` after `continue outer`.

This PR rewrites the benchmark by swapping the `while` and `for-in` loops into more natural order, which also eliminates the need for using `position` at all. This is based on examples from [Rosetta Code](https://www.rosettacode.org/wiki/Roman_numerals/Decode#Swift).

The coverage has been expanded to measure performance of very short string arguments with:
* `String`,
* `Substring`,
* `UTF8View.SubSequence`

and parsing roman numerals to `Int` with appropriate methods variants:

* `hasPrefix`
* `starts(with:)`
* `removeFirst`
* `dropFirst`

and building the roman numeral using `String.append()`.

There are two functional programming style variants of the `Roman.String.hasPrefix.dropFirst`  benchmark (the fastest variant from the imperative `String`/`Substring` groups) with name suffixes: 
* `.R` for `reduce` and
* `.RI` for `reduce(into:)` variants.

The `Roman.Substring.startsWith.dropFirst` variant tests the same code paths as the original `RomanNumbers` benchmark, but the name is changed because of the bug-fix and lowered workload (to run in less than 1000 μs).

For comparison, there's one extra variant with character based parsing algorithm: `Roman.DictCharInt.map.reduce`.

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci smoke test OS X

palimondo commented:
@swift-ci please smoke test OS X

milseman commented:
> Michael, are there any plans to improve Substring performance for small strings?

Some operations will always be a tiny bit slower for small Substrings, because you have to account for the slice's begin/end on top of the small content's count. But, in the future we can use vector loads and extracts for much better improvements, which would help small String and Substring and close any gaps.

I've also seen overhead not getting eliminated from Slice, which is all Substring really is, even when it could be constant folded. CC @eeckstein for that one. Mutations on a Substring could also trigger a COW if a reference to the original string is still live.

palimondo commented:
@milseman Thanks for the insight!

The slower performance of mutable `removeFirst` vs `dropFirst`+`String.init` was on `String` – not `Substring`. I find it strange that forcing a new String copy after going through `Substring` is faster than modifying SmallString in place.

BTW, I’ve locally ran a modified version of these extended benchmarks that were comparing Roman numerals of length 15 and 16 to get a feel for the perf improvement `_SmallString` guts bring and the relative ratios between the different method combinations held there too… Do you agree with Ben that extending the benchmark coverage as demonstrated is not beneficial?

palimondo commented:
@eeckstein Can you please review this? 🙏 It looks like Ben is busy...

palimondo commented:
@eeckstein Thanks for the review. Did you notice us name dropping you upthread about some perf optimization issues, too?

shajrawi commented:
this change broke the build:
```
******************** TEST 'Swift(macosx-x86_64) :: benchmark/Benchmark_O.test.md' FAILED ********************
Script:
--
: 'RUN: at line 26';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md                       --check-prefix LISTPRECOMMIT                       --check-prefix LISTTAGS
: 'RUN: at line 42';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list --skip-tags= | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix LISTALL
: 'RUN: at line 57';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O HashQuadratic --list | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NAMEDSKIP
: 'RUN: at line 60';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O 1 --list | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix RUNBYNUMBER
: 'RUN: at line 67';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list --tags=Dictionary,Array              | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix ANDTAGS
: 'RUN: at line 73';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list --tags=algorithm --skip-tags=validation              | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix TAGSANDSKIPTAGS
: 'RUN: at line 80';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list --tags=algorithm               --skip-tags=validation,Dictionary,String              | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix ORSKIPTAGS
: 'RUN: at line 92';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list              | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix ALPHASORT
: 'RUN: at line 107';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O AngryPhonebook --num-iters=1               | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NUMITERS1                               --check-prefix LOGHEADER                               --check-prefix LOGBENCH
: 'RUN: at line 126';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O 0 --quantile=4 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix FIVENUMSUMMARY
: 'RUN: at line 128';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O 0 --quantile=20 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix VENTILES
: 'RUN: at line 145';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O 1 Ackermann 1 AngryPhonebook               --verbose --num-samples=2 --memory --quantile=2 --delta               | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix RUNJUSTONCE                               --check-prefix CONFIG                               --check-prefix LOGVERBOSE                               --check-prefix MEASUREENV                               --check-prefix LOGFORMAT
: 'RUN: at line 176';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --list --skip-tags= --delim=$'\t' --verbose               | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix LOGVERBOSEDELIM
: 'RUN: at line 185';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --bogus 2>&1               | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix ARGPARSE
: 'RUN: at line 189';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --sample-time          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NOVALUE
: 'RUN: at line 193';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --sample-time=          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix EMPTYVAL
: 'RUN: at line 197';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --sample-time=NaN          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NANVALUE
: 'RUN: at line 201';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --num-iters          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NUMITERS
: 'RUN: at line 205';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --num-samples          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix NUMSAMPLES
: 'RUN: at line 209';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --sleep          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix SLEEP
: 'RUN: at line 213';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --delim          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix DELIM
: 'RUN: at line 217';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --tags=bogus          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix BADTAG
: 'RUN: at line 221';   not /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --skip-tags=bogus          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix BADSKIPTAG
: 'RUN: at line 230';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O 1 2 --memory --list          2>&1 | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix WARNMEMORY
: 'RUN: at line 238';   /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/Benchmark_O --help | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md --check-prefix OPTIONS
--
Exit Code: 1

Command Output (stderr):
--
/Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/test/benchmark/Benchmark_O.test.md:78:18: error: TAGSANDSKIPTAGS: expected string not found in input
TAGSANDSKIPTAGS: RomanNumbers
                 ^
<stdin>:6:14: note: scanning from here
353,Fibonacci,[algorithm, unstable]
             ^
<stdin>:9:5: note: possible intended match here
547,Roman.Substring.startsWith.dropFirst,[String, algorithm, api]
    ^
```

palimondo commented:
@shajrawi I'm sorry about that. Thank's for the quick revert. I didn't realize the `lit` test depended on that name. I'll do this correctly on Monday. My apologies!

milseman commented:
@palimondo testing small and large workloads makes perfect sense to me.

Any difference in perf between removeFirst and dropFirst+init should be filed as a bug, but probably not tracked as separate benchmarks. Ideally, it'd be a LIT test. @eeckstein or others would have more insight there

palimondo commented:
@milseman Regarding the benchmark vs. LIT tests, I would agree in principle, but I think its better to add benchmarks first to highlight performance issues, then file performance bugs against them and finally remove the "duplicates" (identically performing benchmarks) once the fix is in place and verified by the LIT test. In other words: as a first step I'm proposing to significantly increase performance coverage (add lot more benchmarks, mainly to showcase oddly regressive variants) and these should be replaced step-by-step by LIT tests as the individual regressions get fixed.

This starts to feel like something that should be discussed more broadly in the forums? (I had similar debates with Andrew, Erik and others scattered in various PRs…)
